internal/auth/configuration.go

```diff
@@ -0,0 +1,437 @@
+			return xerrors.Errorf("invalid provider.okta config: %w", err)
+		}
+	case "test":
+		break
+	default:
+		return xerrors.Errorf("unknown provider.type: %q", pc.ProviderType)
+	}
+
+	if err := pc.GroupCacheConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid provider.groupcache config: %w", err)
+	}
+
+	return nil
+}
+
+type GoogleProviderConfig struct {
+	Credentials    string `mapstructure:"credentials"`
+	Impersonate    string `mapstructure:"impersonate"`
+	ApprovalPrompt string `mapstructure:"prompt"`
+}
+
+func (gpc GoogleProviderConfig) Validate() error {
+	// TOOD: what can we validate here?
+	return nil
+}
+
+type OktaProviderConfig struct {
+	ServerID string `mapstructure:"server"`
+	OrgURL   string `mapstructure:"url"`
+}
+
+func (opc OktaProviderConfig) Validate() error {
+	if opc.OrgURL == "" {
+		return xerrors.New("no okta.url is configured")
+	}
+
+	if opc.ServerID == "" {
+		return xerrors.New("no okta.server is configured")
+	}
+
+	return nil
+}
+
+type GroupCacheConfig struct {
+	Interval struct {
+		Provider time.Duration `mapstructure:"provider"`
+		Refresh  time.Duration `mapstructure:"refresh"`
+	} `mapstructure:"interval"`
+}
+
+func (gcc GroupCacheConfig) Validate() error {
+	return nil
+}
+
+type SessionConfig struct {
+	CookieConfig CookieConfig `mapstructure:"cookie"`
+
+	SessionLifetimeTTL time.Duration `mapstructure:"lifetime"`
+	Key                string        `mapstructure:"key"`
+}
+
+func (sc SessionConfig) Validate() error {
+	// TODO: we can validate this secret is usable
+	if sc.Key == "" {
+		return xerrors.New("no session.key configured")
+	}
+
+	if _, err := base64.StdEncoding.DecodeString(sc.Key); err != nil {
+		return xerrors.Errorf("invalid session.key: %w", err)
+	}
+
+	if sc.SessionLifetimeTTL >= (365*24)*time.Hour || sc.SessionLifetimeTTL <= 1*time.Minute {
+		return xerrors.Errorf("session.lifetime must be between 1 minute and 1 year but is: %v", sc.SessionLifetimeTTL)
+	}
+
+	if err := sc.CookieConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid session.cookie config: %w", err)
+	}
+
+	return nil
+}
+
+type CookieConfig struct {
+	Name     string        `mapstructure:"name"`
+	Secret   string        `mapstructure:"secret"`
+	Domain   string        `mapstructure:"domain"`
+	Expire   time.Duration `mapstructure:"expire"`
+	Secure   bool          `mapstructure:"secure"`
+	HTTPOnly bool          `mapstructure:"httponly"`
+}
+
+func (cc CookieConfig) Validate() error {
+	// TODO: Validate cookie secret:
+	if cc.Name == "" {
+		return xerrors.New("no cookie.name configured")
+	}
+
+	cookie := &http.Cookie{Name: cc.Name}
+	if cookie.String() == "" {
+		return xerrors.Errorf("invalid cookie.name: %q", cc.Name)
+	}
+
+	if cc.Secret == "" {
+		return xerrors.New("no cookie.secret configured")
+	}
+
+	if _, err := base64.StdEncoding.DecodeString(cc.Secret); err != nil {
+		return xerrors.Errorf("invalid cookie.secret: %w", err)
```

Comment:
Upon error we previously would also output how this variable _should_ be created - I don't think it necessarily needs to be added to the output, but I just want to make sure that information isn't lost as I know it's something that was brought up in the past.